### PR TITLE
Add support for KerasSpiking layers

### PR DESCRIPTION
**Pull request by drasmuss**
Also added support for TimeDistributed layers, since that's present in a lot of KerasSpiking models. And then added some pointers to KerasSpiking in various places throughout the documentation.

### CODE REVIEW

File: nengo_dl/tests/test_converter.py [hunk 1]
```diff
@@ -809,3 +820,182 @@ def relu2(x):
...
+    tf.keras.backend.set_value(
+        layer.layer.cell.tau_var, np.arange(10).reshape(layer.layer.cell.tau_var.shape)
+    )
+    model = tf.keras.Model(inp, layer(inp))
+    with pytest.raises(TypeError, match="different tau values"):
+        converter.Converter(model, temporal_model=True)
+
+
+def test_time_distributed():
+    inp = x = tf.keras.Input((10, 8))
```
Comment (hunse) [inline nengo_dl/tests/test_converter.py:989]: How much work would it be to also support `None` for the number of timesteps?

Comment (hunse) [reply nengo_dl/tests/test_converter.py:989]: Never mind, I realized this (seems to) work, I was just getting an exception in the second part of this test when I switched it.

File: nengo_dl/converter.py [hunk 2]
```diff
@@ -84,6 +84,11 @@ class Converter:
         increasing the synaptic filtering will make the network less responsive to
         rapid changes in the input, and you may need to present each input value
         for more timesteps in order to allow the network output to settle.
+    temporal_model : bool
+        Set to True if the Keras model contains a temporal
+        dimension (e.g. the inputs and outputs of each layer have shape
+        ``(batch_size, n_steps, ...)``). Note that all layers must be temporal, the
+        Converter cannot handle models with a mix of temporal and non-temporal layers.
```
Comment (hunse) [inline nengo_dl/converter.py:91]: I'm trying to think of ways we could get around this for at least some of the simple/more common cases. It's pretty common for models to collapse over time in some way near the end.

What if rather than removing the first two dimensions for all layers, we only do that if `rank > 2`. So any layer that's already flattened would be treated as not having time. That would at least allow e.g. a dense layer at the end of the model that does the classification.

Comment (hunse) [reply nengo_dl/converter.py:91]: Never mind. While the shape trick is possible, there's no way to pool across time in Nengo, since the whole network has to run each timestep.

Probably the best way to do something equivalent is to have time in all Keras layers, but then have a filter near the end of the network, and a loss function that just uses the last timestep. We could automate this "use the last timestep in the loss function" if e.g. we detect a filter layer with `return_sequences=False`, but it's not as trivial as I thought (and probably somewhat fragile, too).

**Review by hunse — APPROVED** [review]

File: nengo_dl/converter.py [hunk 3]
```diff
@@ -1450,7 +1450,11 @@ class ConvertInput(LayerConverter):
     def convert(self, node_id):
         shape = self.output_shape(node_id)
         if any(x is None for x in shape):
-            raise ValueError("Input shapes must be fully specified; got %s" % (shape,))
+            raise ValueError(
+                "Input shapes must be fully specified; got %s. For inputs with `None` "
+                "in the first axis, indicating a variable number of timesteps, "
```
[the remark this answers is not in the document]
Comment (drasmuss) [reply nengo_dl/converter.py:1455]: Oh nm I was parsing this wrong, this is just helping to give them a hint that they should set `temporal_model=True`. I might rephrase to something like "If inputs contain `None` in the first axis to indicate a variable number of timesteps, ..." (I read the current phrasing as implying that if the first axis is None that means it is a temporal model, which isn't necessarily True).

File: nengo_dl/tests/test_converter.py [hunk 4]
```diff
@@ -995,7 +996,8 @@ def test_time_distributed():
 
     _test_convert(inp, x, temporal_model=True)
 
-    with pytest.raises(
-        TypeError, match="can only be converted when temporal_model=True"
-    ):
-        converter.Converter(tf.keras.Model(inp, x), temporal_model=False)
+    if timesteps is not None:
```
Comment (drasmuss) [inline nengo_dl/tests/test_converter.py:1009]: This error should be raised in both cases I think

Comment (hunse) [reply nengo_dl/tests/test_converter.py:1009]: No, I think if `timesteps is None`, then it's the `ValueError` in `ConvertInput` that we were discussing above. That's why I added that comment about setting `temporal_model=True` there, since if `timesteps is None` and `temporal_model=False`, that's the first error the user sees. We could test that other error here, too, though.

Comment (drasmuss) [reply nengo_dl/tests/test_converter.py:1009]: Ah yeah that's true I guess there isn't an obvious way to hit this in the `timesteps=None` case. I might catch the other error here, just to make it clear that it's an error in both cases (so no one reading this test thinks it should work in the other case, like my mistake).

**Comment by hunse** [comment]
Ok, made your suggested changes @drasmuss

**Comment by drasmuss** [comment]
Fixups look good
